agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -1167,24 +1168,52 @@ func (dg *dockerGoClient) Stats(id string, ctx context.Context) (<-chan *docker.
 		return nil, err
 	}
 
-	stats := make(chan *docker.Stats)
-	options := docker.StatsOptions{
-		ID:                id,
-		Stats:             stats,
-		Stream:            true,
-		Context:           ctx,
-		InactivityTimeout: StatsInactivityTimeout,
-	}
+	returnStats := make(chan *docker.Stats)
 
-	go func() {
-		statsErr := client.Stats(options)
-		if statsErr != nil {
-			seelog.Infof("DockerGoClient: Unable to retrieve stats for container %s: %v",
-				id, statsErr)
+	if !dg.config.PollMetrics {
+		seelog.Infof("DockerGoClient: Starting to Stream for metrics")
+		options := docker.StatsOptions{
+			ID:                id,
+			Stats:             returnStats,
+			Stream:            true,
+			Context:           ctx,
+			InactivityTimeout: StatsInactivityTimeout,
 		}
-	}()
+	
+		go func() {
+			statsErr := client.Stats(options)
+			if statsErr != nil {
+				seelog.Infof("DockerGoClient: Unable to retrieve stats for container %s: %v",
+					id, statsErr)
+			}
+		}()
+	} else {
+		seelog.Infof("DockerGoClient: Starting to Poll for metrics")
+		//Sleep for a random period time up to the polling interval. This will help make containers ask for stats at different times
+		time.Sleep(time.Second * time.Duration(rand.Intn(int(dg.config.PollingMetricsWaitDuration.Seconds()))))
```

Comment:
Good intuition but I don't think this is needed. The only time all containers of a Task would poll at the same time is if Agent stops and starts while the Task is running. Otherwise, assuming go routines run at approximately the same speed (or even randomly), containers are added to the stats engine iteratively so they will already be staggered to some degree.

Reply:
I think there is value in adding extra jitter here.